### DISCUSSION TITLE
SSG - Questions about rendering workflow and recommandation in regards to Analytics

### DISCUSSION THREAD

**Discussion by Vadorequest** [vercel/next.js#12451] [Help]
I'm trying to configure Amplitude on a page with SSG enabled.

```tsx
    <AmplitudeProvider
      amplitudeInstance={amplitudeInstance}
      apiKey={process.env.AMPLITUDE_API_KEY}
      userId={userId}
    >
      <Amplitude
        // DA Event props and user props are sometimes duplicated to ease the data analysis through Amplitude
        //  Because charts are sometimes easier to build using events props, or user users props
        eventProperties={{
          app: {
            name: process.env.APP_NAME,
            version: process.env.APP_VERSION,
            stage: process.env.APP_STAGE,
          },
          page: {
            url: location.href,
            path: location.pathname,
            origin: location.origin,
            name: null,
          },
          customer: {
            ref: customerRef,
          },
          lang: lang,
          locale: locale,
          iframe: isInIframe,
          iframeReferrer: iframeReferrer,
        }}
        // userProperties={{}} XXX Do not use this, add default user-related properties in initAmplitude instead
      >
        some code
      </Amplitude>
    </AmplitudeProvider
```

This fails with `ReferenceError: location is not defined`. I could fix it quite easily using `location?.href`, but I wonder about a few things first.

1. Is there an **official way** to know if the page is currently built by the server (static generation) or being executed at runtime? I could check for the existence of `window` in my page `render`, but I wonder what's the official recommandation for doing this.
1. How is it recommended to do stuff related to `location` or `browser-related` (cookies, localStorage)? What's the rendering workflow with SSG? My main concern would be the app rendering twice (once using the static markup, and then again with the actual data)
    - I believe it only renders once, and I'll try it ASAP, but I'm still wondering about official recommandations and advices.
1. How should be treated errors about **mismatch between server and browser markup**? 
    - e.g: `Warning: Text content did not match. Server: "cd563de0-8d76-11ea-94c7-7b2a228dc097" Client: "a234bd00-8d65-11ea-a032-a3f33ca380d6"`, I'm generating an id for the user if none exists (stored in cookies) but obviously the generated id isn't the same between the server render (no cookie, new ID is created) and browser render (cookie exists, use existing id). I believe this only happens during development, and the warning is misleading because it makes believe something is wrong, but I don't see how the server and browser could yield the same output in such case.

**Comment by brianlovin** [comment]
2. Anything client-related should happen after mount, usually inside some kind of `useEffect` call. Assume that everything else is being computed on the server or at build time.
3. I'd recommend looking into https://joshwcomeau.com/react/the-perils-of-rehydration/ -> this is a great walkthrough of why mismatches happen and how to think systematically about fixing them. If you're rending content based on information in cookies or localStorage, and also doing SSG, you will keep hitting mismatches. One solution to fix this is to _always_ render an empty / placeholder value to account for SSG, and refresh the value only when the component mounts on the client and has window access.

**Reply by Vadorequest** [reply]
Thank you! I'll read it through, looks very interesting indeed!

Here is a demo of the app I'm building, it's fairly simple. Not the `Device id` in the footer (bottom), which comes from browser cookies:
https://nrn-v1-hyb-mst-aptd-gcms-lcz-sty-c1-1w6t2bwrm.now.sh/

I have changed my implementation and I've naturally followed the "render empty on !browser" rule. It's invisible to the human eye. 

Before, I used to init my cookies from the server side, which obviously created an `id` mismatch. I agree that using empty content is cleaner, and also forces developers to avoid doing browser-related stuff when building the static content.

**Reply by brianlovin** [reply]
This might be a useful hook, as well: https://joshwcomeau.com/snippets/react-hooks/use-has-mounted - it's slightly different from rendering `null`, but instead focuses on the mounting itself.

**Reply by Vadorequest** [reply]
https://joshwcomeau.com/react/the-perils-of-rehydration/ really answered most of my questions and even more, great resource. Thanks for bringing it up!

I understand better how things work, I wasn't so familiar with rehydration and how to handle it properly. I now understand that ignoring the `Text content did not match` warning wasn't really helpful.
